[lldb] Introduce ScriptedFrame affordance

medismailben opened the pull request:
This patch introduces a new scripting affordance in lldb: `ScriptedFrame`.

This allows user to produce mock stackframes in scripted threads and
scripted processes from a python script.

With this change, StackFrame can be synthetized from different sources:
- Either from a dictionary containing a load address, and a frame index,
  which is the legacy way.
- Or by creating a ScriptedFrame python object.

One particularity of synthezising stackframes from the ScriptedFrame
python object, is that these frame have an optional PC, meaning that
they don't have a report a valid PC and they can act as shells that just
contain static information, like the frame function name, the list of
variables or registers, etc. It can also provide a symbol context.

rdar://157260006

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

jimingham commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+  if (!thread.IsValid())
+    return llvm::createStringError("Invalid scripted thread.");
+
+  thread.CheckInterpreterAndScriptObject();
```
If the asserts in CheckInterpreterAndScriptObject fail, do you really want to continue?  Also, I think you meant `Interface` in the method name and not `Interpreter`, since that's what you actually check...

jimingham commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+  thread.CheckInterpreterAndScriptObject();
+
+  auto scripted_frame_interface =
+      thread.GetInterface()->CreateScriptedFrameInterface();
```
Yes, if the check failed, this would be unsafe.

jimingham commented on lldb/source/Plugins/Process/scripted/ScriptedThread.cpp:
```diff
@@ -204,9 +209,64 @@ bool ScriptedThread::LoadArtificialStackFrames() {
+  for (size_t idx = 0; idx < arr_size; idx++) {
+    StackFrameSP synth_frame_sp = nullptr;
+
+    auto frame_from_dict_or_err = create_frame_from_dict(idx);
```
This doesn't seem right.  There are two reasons create_frame_from_dict could fail:
1) Because the frame isn't described by a dictionary but by a script object
2) Because the dict existed but was incorrect
In case 1) you do want to try the scripted object instead, but not if there was a dictionary but it failed.  Falling back to the script object and obscuring the fact that there was a bogus dictionary will be confusing.

jimingham replied:
I would expect that if there IS a dictionary provider that returns an actual error handling that index, we should report the error.  People would want to know that there was this dictionary provider floating around possibly interfering with the synthetic provider they have just added...
If the dictionary provider says it isn't in the business of providing that frame, that shouldn't be an error since that just means "that index is not my job".  I presume it would do that by returning an empty frame but not an error?
If the dictionary provider says "that index wasn't something I wanted to provide a frame for", you should go on and consult the synthetic one.
Also, if we are forced to handle the case where there are both synthetic and dictionary providers at the same time, we should do the synthetic one first.  After all, since that's the newer feature, it makes more sense to let it override the older variant than the other way round...

jimingham replied:
Right, the main thing that was bugging me was if this bit:

```
    if (!dict->GetValueForKeyAsInteger("pc", pc)) {
      ScriptedInterface::ErrorWithMessage<bool>(
          LLVM_PRETTY_FUNCTION,
          "Couldn't find value for key 'pc' in stackframe dictionary.", error,
           LLDBLog::Thread);
      return error.ToError();
    }
```

fails, you really don't want to do more work.  After all, this ScriptedThread implementation provided you what it thought was a legit dictionary entry for a frame, but it was in fact bogus.  In that case, you shouldn't try to do any more work, you should just report the error.

Internally there's just one array that either has the dictionary or the script object, so you won't lose the error.  But you also won't succeed in getting the script object out of it either because you already found a Dictionary entry there.

So the code would be cleaner if you made a distinction between "there's something other than a Dictionary in this slot" and "there's a bad Dictionary in this slot".

And even if we changed the implementation in the future so that the Dictionaries and the Script Objects didn't live in the same array, this would continue to be the right behavior.  If the ScriptedThread claims to return a Dictionary representation for that index but it isn't correct, we should report its badness rather than trying the Scripted Frame version.

jimingham commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedFramePythonInterface.cpp:
```diff
@@ -0,0 +1,158 @@
+
+  if (!ScriptedInterface::CheckStructuredDataObject(LLVM_PRETTY_FUNCTION, obj,
+                                                    error))
+    return {};
```
What is {} for a bool?

medismailben replied:
false

jimingham replied:
Yeah, just return false, that's a dopey thing to have to know...

medismailben commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+ScriptedFrame::~ScriptedFrame() {}
+
+const char *ScriptedFrame::GetFunctionName() {
+  CheckInterpreterAndScriptObject();
```
ditto

medismailben commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+  return ConstString(function_name->c_str()).AsCString();
+}
+
+const char *ScriptedFrame::GetDisplayFunctionName() {
```
Well, this is just the current return type in the base `lldb_private::StackFrame` class.  This change seems unrelated to this PR however.

JDevlieghere replied:
Agreed that it should be a separate PR, but I think it's worth doing first instead of making things worse by going through ConstString.

JDevlieghere commented on lldb/include/lldb/Interpreter/Interfaces/ScriptedFrameInterface.h:
```diff
@@ -0,0 +1,57 @@
+//===-- ScriptedFrameInterface.h --------------------------------*- C++ -*-===//
```
```suggestion
//===----------------------------------------------------------------------===//
```

JDevlieghere commented on lldb/include/lldb/Interpreter/Interfaces/ScriptedFrameInterface.h:
```diff
@@ -0,0 +1,57 @@
+#ifndef LLDB_INTERPRETER_INTERFACES_SCRIPTEDFRAMEINTERFACE_H
+#define LLDB_INTERPRETER_INTERFACES_SCRIPTEDFRAMEINTERFACE_H
+
+#include "ScriptedInterface.h"
+#include "lldb/Core/StructuredDataImpl.h"
+#include "lldb/Symbol/SymbolContext.h"
+
+#include "lldb/lldb-private.h"
+
+#include <optional>
+#include <string>
```
```suggestion
#include "ScriptedInterface.h"
#include "lldb/Core/StructuredDataImpl.h"
#include "lldb/Symbol/SymbolContext.h"
#include "lldb/lldb-private.h"
#include <optional>
#include <string>
```

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+//===-- ScriptedFrame.cpp -------------------------------------------------===//
```
```suggestion
//===----------------------------------------------------------------------===//
```

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+      thread.GetInterface()->CreateScriptedFrameInterface();
+  if (!scripted_frame_interface)
+    return llvm::createStringError(
+        "Failed to create scripted frame interface.");
```
```suggestion
        "failed to create scripted frame interface");
```
Error message should not be capitalized and have no period. Applies across this file.

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+      frame_class_name, exe_ctx, args_sp, script_object);
+
+  if (!obj_or_err) {
+    llvm::consumeError(obj_or_err.takeError());
+    return llvm::createStringError("Failed to create script object.");
```
?

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+  if (!owned_script_object_sp->IsValid())
+    return llvm::createStringError("Created script object is invalid.");
+
+  lldb::user_id_t frame_id = scripted_frame_interface->GetID();
```
Can you move this closer to where it is used? Make it `const` if it's not used.

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+
+  std::optional<SymbolContext> maybe_sym_ctx =
+      scripted_frame_interface->GetSymbolContext();
+  if (maybe_sym_ctx) {
+    sc = *maybe_sym_ctx;
+  }
```
```suggestion
  if (maybe_sym_ctx) 
    sc = *maybe_sym_ctx;
```

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+  std::optional<std::string> reg_data =
+      scripted_frame_interface->GetRegisterContext();
+  if (reg_data) {
+    DataBufferSP data_sp(
+        std::make_shared<DataBufferHeap>(reg_data->c_str(), reg_data->size()));
```
```suggestion
    DataBufferSP data_sp = 
        std::make_shared<DataBufferHeap>(reg_data->c_str(), reg_data->size());
```

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -0,0 +1,192 @@
+    if (!data_sp->GetByteSize())
+      return llvm::createStringError("Failed to copy raw registers data");
+
+    std::shared_ptr<RegisterContextMemory> reg_ctx_memory =
+        std::make_shared<RegisterContextMemory>(
+            thread, frame_id, *register_info_sp, LLDB_INVALID_ADDRESS);
```
```suggestion
    auto reg_ctx_memory =
        std::make_shared<RegisterContextMemory>(
            thread, frame_id, *register_info_sp, LLDB_INVALID_ADDRESS);
```

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.h:
```diff
@@ -0,0 +1,65 @@
+//===-- ScriptedFrame.h -----------------------------------------*- C++ -*-===//
```
```suggestion
//===----------------------------------------------------------------------===//
```

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.h:
```diff
@@ -0,0 +1,65 @@
+#ifndef LLDB_SOURCE_PLUGINS_SCRIPTED_FRAME_H
+#define LLDB_SOURCE_PLUGINS_SCRIPTED_FRAME_H
+
+#include <string>
+
+#include "ScriptedThread.h"
+
+#include "Plugins/Process/Utility/RegisterContextMemory.h"
+#include "lldb/Interpreter/ScriptInterpreter.h"
+#include "lldb/Target/DynamicRegisterInfo.h"
+#include "lldb/Target/StackFrame.h"
```
The include order is inverted here. Leave out the newline and clang-format do it right for you. 
```suggestion
#include <string>
#include "ScriptedThread.h"
#include "Plugins/Process/Utility/RegisterContextMemory.h"
#include "lldb/Interpreter/ScriptInterpreter.h"
#include "lldb/Target/DynamicRegisterInfo.h"
#include "lldb/Target/StackFrame.h"
```

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.h:
```diff
@@ -0,0 +1,65 @@
+
+  std::shared_ptr<DynamicRegisterInfo> GetDynamicRegisterInfo();
+
+  lldb::ScriptedFrameInterfaceSP m_scripted_frame_interface_sp = nullptr;
+  lldb_private::StructuredData::GenericSP m_script_object_sp = nullptr;
+  std::shared_ptr<DynamicRegisterInfo> m_register_info_sp = nullptr;
```
These are already default initialized to nullptr. 
```suggestion
  lldb::ScriptedFrameInterfaceSP m_scripted_frame_interface_sp;
  lldb_private::StructuredData::GenericSP m_script_object_sp;
  std::shared_ptr<DynamicRegisterInfo> m_register_info_sp;
```

JDevlieghere commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedFramePythonInterface.cpp:
```diff
@@ -0,0 +1,157 @@
+//===-- ScriptedFramePythonInterface.cpp ----------------------------------===//
```
```suggestion
//===----------------------------------------------------------------------===//
```

JDevlieghere commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedFramePythonInterface.h:
```diff
@@ -0,0 +1,59 @@
+//===-- ScriptedFramePythonInterface.h --------------------------*- C++ -*-===//
```
```suggestion
//===----------------------------------------------------------------------===//
```

JDevlieghere commented on lldb/test/API/functionalities/scripted_process/dummy_scripted_process.py:
```diff
@@ -109,6 +113,61 @@ def get_register_context(self) -> str:
+
+            return sym_ctx_list.GetContextAtIndex(0)
+
+        return get_symbol_context_for_function(self.name)
```
Can you add a test with a complete made up symbol context?

JDevlieghere commented on lldb/bindings/python/python-wrapper.swig:
```diff
@@ -519,6 +519,7 @@ void *lldb_private::python::LLDBSWIGPython_CastPyObjectToSBMemoryRegionInfo(PyOb
   return sb_ptr;
 }
 
+
```
Spurious newline.

JDevlieghere reviewed (approved):
LGTM

medismailben committed:
[lldb] Introduce ScriptedFrame affordance

This patch introduces a new scripting affordance in lldb:  `ScriptedFrame`.

This allows user to produce mock stackframes in scripted threads and
scripted processes from a python script.

With this change, StackFrame can be synthetized from different sources:
- Either from a dictionary containing a load address, and a frame index,
  which is the legacy way.
- Or by creating a ScriptedFrame python object.

One particularity of synthezising stackframes from the ScriptedFrame
python object, is that these frame have an optional PC, meaning that
they don't have a report a valid PC and they can act as shells that just
contain static information, like the frame function name, the list of
variables or registers, etc. It can also provide a symbol context.

rdar://157260006

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>
